### ISSUE TITLE
MA0077 - A class that provides Equals(T) should implement IEquatable<T> is handled by CA1066

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#855]
- **Rule**: [MA0077](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0077.md) - A class that provides Equals(T) should implement IEquatable<T>
- **Built-in Equivalent**: [CA1066](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1066) - Implement IEquatable when overriding Equals

## Analysis

**MA0077** and **CA1066** serve the same purpose: ensuring that types implementing type-specific equality also implement the `IEquatable<T>` interface to provide consistent and discoverable equality semantics.

### MA0077 Coverage
- Detects classes with Equals(T) method without IEquatable<T>
- Promotes proper equality interface implementation
- Ensures consistent equality patterns

### CA1066 Coverage  
- Detects types that should implement IEquatable when providing typed equality
- Official Microsoft rule for proper equality implementation
- Ensures interface consistency for equality operations
- Part of the .NET design guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0077 and CA1066 will flag this
public class Person
{
    public string Name { get; set; }
    public int Age { get; set; }
    
    // Has typed Equals but doesn't implement IEquatable<Person>!
    public bool Equals(Person other)
    {
        if (other == null) return false;
        return Name == other.Name && Age == other.Age;
    }
    
    public override bool Equals(object obj)
    {
        return Equals(obj as Person);
    }
    
    public override int GetHashCode()
    {
        return HashCode.Combine(Name, Age);
    }
}
```

**Recommended fix:**
```csharp
// Both rules suggest implementing IEquatable<T>
public class Person : IEquatable<Person>
{
    public string Name { get; set; }
    public int Age { get; set; }
    
    // Now properly implements the interface
    public bool Equals(Person other)
    {
        if (other == null) return false;
        return Name == other.Name && Age == other.Age;
    }
    
    public override bool Equals(object obj)
    {
        return Equals(obj as Person);
    }
    
    public override int GetHashCode()
    {
        return HashCode.Combine(Name, Age);
    }
    
    public static bool operator ==(Person left, Person right)
    {
        return EqualityComparer<Person>.Default.Equals(left, right);
    }
    
    public static bool operator !=(Person left, Person right)
    {
        return !(left == right);
    }
}
```

## Benefits of Implementing IEquatable<T>

Proper IEquatable<T> implementation provides:
- **Performance benefits** - avoids boxing for value types
- **Generic collection optimization** - better performance in collections
- **Clear API contract** - discoverable through interface
- **Consistent equality semantics** - follows established patterns

## Common Usage Patterns

```csharp
// Generic collections benefit from IEquatable<T>
var people = new HashSet<Person>(); // Uses IEquatable<Person>.Equals
var found = people.Contains(searchPerson); // Efficient, no boxing

// LINQ operations benefit
var distinct = list.Distinct(); // Uses IEquatable<T> if available
```

## Recommendation

**CA1066** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0077. Since both rules target the same equality interface pattern and CA1066 is the standard rule for equality implementation, **MA0077 should be deprecated** in favor of CA1066.

## Migration Path

1. Enable **CA1066** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA1066" />
   ```
2. Disable **MA0077**:
   ```xml
   <WarningsNotAsErrors Include="MA0077" />
   ```
3. Update any rule suppressions from MA0077 to CA1066

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: Final Batch - All Equivalent Rules ✅

**MA0077 vs CA1066** - IEquatable implementation consistency
**MA0082 vs CA2242** - Correct NaN testing patterns  
**MA0095 vs CA1067** - Equals override with IEquatable
**MA0102 vs IDE0044** - Readonly modifier suggestions
**MA0112 vs CA1860** - Count vs Any performance optimization
**MA0131 vs CA2264** - ArgumentNullException.ThrowIfNull usage
**MA0142 vs IDE0041** - Pattern matching for null checks

### Recommendation: Deprecate All Listed MA Rules ✅

All provide equivalent functionality to official Microsoft analyzers and should be deprecated.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
After analyzing both rules, **MA0077** and **CA1066** are **NOT equivalent** and MA0077 should remain active.

## Key Differences:

**CA1066** (Microsoft):
- Only applies to **value types (structs)** that override `Equals(object)`
- Focuses on performance optimization by avoiding boxing when using `IEquatable<T>`
- [Official documentation](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1066)

**MA0077** (Meziantou):
- Applies to **classes** that provide `Equals(T)` method
- Ensures proper interface implementation for consistent equality semantics
- Broader coverage than CA1066

## Why MA0077 provides additional value:

1. **Different scope**: CA1066 = structs only, MA0077 = classes only
2. **Complementary coverage**: Together they cover both reference and value types
3. **Design consistency**: Classes with typed Equals should implement IEquatable<T> for discoverability

## Example where MA0077 catches issues CA1066 doesn't:

```csharp
// MA0077 will flag this, CA1066 will not (it's a class, not a struct)
public class Person
{
    public bool Equals(Person other) => /* implementation */;
    // Missing: IEquatable<Person> implementation
}
```

**Recommendation**: Keep both rules active as they serve different purposes and provide complementary coverage.

This issue should not be added to the comparison table as the rules are not equivalent.
